charger/mcc.go

```diff
@@ -0,0 +1,406 @@
+func (mcc *MCC) getEscapedJSON(uri string, result interface{}) error {
+	req, err := mcc.customRequest(http.MethodGet, uri)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err == nil {
+		var unquote string
+		if err := json.Unmarshal([]byte(b), &unquote); err != nil {
+			return err
+		}
+
+		if err := json.Unmarshal([]byte(unquote), &result); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// Status implements the Charger.Status interface
+func (mcc *MCC) Status() (api.ChargeStatus, error) {
+	var chargeState int64
+
+	err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState)
+
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	// Charge State values and mapping to ChargeStatus
+	// 0: Unplugged		StatusA
+	// 1: Connecting	StatusB
+	// 2: Error				StatusE
+	// 3: Established	StatusF
+	// 4: Paused			StatusD
+	// 5: Active			StatusC
+	// 6: Finished		StatusD
+
+	switch chargeState {
+	case 0:
+		return api.StatusA, nil
+	case 1:
+		return api.StatusB, nil
+	case 2:
+		return api.StatusE, nil
+	case 3:
+		return api.StatusF, nil
+	case 4, 6:
+		return api.StatusD, nil
+	case 5:
+		return api.StatusC, nil
+	default:
+		return api.StatusNone, fmt.Errorf("properties unknown result: %d", chargeState)
+	}
+}
+
+// Enabled implements the Charger.Enabled interface
+func (mcc *MCC) Enabled() (bool, error) {
+	// Check if the car is connected and Paused, Active, or Finished
+	var chargeState int64
+
+	err := mcc.getValue(mcc.apiURL(apiChargeState), &chargeState)
+
+	if err != nil {
+		return false, err
+	}
+
+	if chargeState >= 4 && chargeState <= 6 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Enable implements the Charger.Enable interface
+func (mcc *MCC) Enable(enable bool) error {
+	// As we don't know of the API to disable charging this for now always returns an error
+
+	return fmt.Errorf("The device doesn't allow to enable/disable the device via an API")
+}
+
+// MaxCurrent implements the Charger.MaxCurrent interface
+func (mcc *MCC) MaxCurrent(current int64) error {
+	// The device doesn't return an error if we set a value greater than the
+	// current allowed max or smaller than the allowed min
+	// instead it will simply set it to max or min and return "OK" anyway
+	// Since the API here works differently, we fetch the limits
+	// and then return an error if the value is outside of the limits or
+	// otherwise set the new value
+	var cableInformation MCCCurrentCableInformation
+	err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentCableInformation), &cableInformation)
+	if err != nil {
+		return err
+	}
+
+	if current < cableInformation.MinValue {
+		return fmt.Errorf("value is lower than the allowed minimum value %d", cableInformation.MinValue)
+	}
+
+	if current > cableInformation.MaxValue {
+		return fmt.Errorf("value is higher than the allowed maximum value %d", cableInformation.MaxValue)
+	}
+
+	url := fmt.Sprintf("%s%d", mcc.apiURL(apiSetCurrentLimit), current)
+	fmt.Println(url)
```

Comment:
leftover

Reply:
Ups 🤦

Reply:
Du musst auch nicht kommentieren- wenns in neuem commit geändert ist mach gh die commants automatisch zu ;)

Reply:
Ich kommuniziere nun mal gerne 😁